capsules/src/virtual_adc.rs

```diff
@@ -0,0 +1,291 @@
+        let adc_user = AdcDevice {
+            mux: mux,
+            channel: channel,
+            operation: OptionalCell::empty(),
+            next: ListLink::empty(),
+            client: OptionalCell::empty(),
+        };
+        adc_user
+    }
+
+    pub fn add_to_mux(&'a self) {
+        self.mux.devices.push_head(self);
+    }
+}
+
+impl<'a, A: hil::adc::Adc> ListNode<'a, AdcDevice<'a, A>> for AdcDevice<'a, A> {
+    fn next(&'a self) -> &'a ListLink<'a, AdcDevice<'a, A>> {
+        &self.next
+    }
+}
+
+impl<A: hil::adc::Adc> hil::adc::AdcChannel for AdcDevice<'_, A> {
+    fn sample(&self) -> ReturnCode {
+        self.operation.set(Operation::OneSample);
+        self.mux.do_next_op();
+        ReturnCode::SUCCESS
+    }
+
+    fn stop_sampling(&self) -> ReturnCode {
+        self.operation.clear();
+        self.mux.do_next_op();
+        ReturnCode::SUCCESS
+    }
+
+    fn sample_continuous(&self) -> ReturnCode {
+        ReturnCode::ENOSUPPORT
+    }
+
+    fn get_resolution_bits(&self) -> usize {
+        self.mux.get_resolution_bits()
+    }
+
+    fn get_voltage_reference_mv(&self) -> Option<usize> {
+        self.mux.get_voltage_reference_mv()
+    }
+    fn set_client(&self, client: &'static dyn hil::adc::Client) {
+        self.client.set(client);
+    }
+}
+
+impl<'a> AdcSyscall<'a> {
+    pub fn new(drivers: &'a [&'a dyn hil::adc::AdcChannel], grant: Grant<App>) -> AdcSyscall<'a> {
+        AdcSyscall {
+            drivers: drivers,
+            apps: grant,
+            current_app: OptionalCell::empty(),
+        }
+    }
+
+    fn enqueue_command(&self, command: Operation, channel: usize, appid: AppId) -> ReturnCode {
+        if channel < self.drivers.len() {
+            self.apps
+                .enter(appid, |app, _| {
+                    if self.current_app.is_none() {
+                        self.current_app.set(appid);
+                        let value = self.call_driver(command, channel);
+                        if value != ReturnCode::SUCCESS {
+                            self.current_app.clear();
+                        }
+                        value
+                    } else {
+                        if app.pending_command == true {
+                            ReturnCode::EBUSY
+                        } else {
+                            app.pending_command = true;
+                            app.command.set(command);
+                            app.channel = channel;
+                            ReturnCode::SUCCESS
+                        }
+                    }
+                })
+                .unwrap_or_else(|err| err.into())
+        } else {
+            ReturnCode::ENODEVICE
+        }
+    }
+
+    fn call_driver(&self, command: Operation, channel: usize) -> ReturnCode {
+        match command {
+            Operation::OneSample => self.drivers[channel].sample(),
+        }
+    }
+}
+
+impl Driver for AdcSyscall<'_> {
```

Comment:
This is out-of-scope for a `virtual_X.rs` file. A new Driver should be in its own capsule.

Reply:
I will move it.